source/ShipInfoPanel.cpp

```diff
@@ -463,44 +473,152 @@ void ShipInfoPanel::DrawWeapons(const Rectangle &bounds)
 	Color topColor;
 	bool hasTop = false;
 	auto layout = Layout(static_cast<int>(LABEL_WIDTH), Truncate::BACK);
+
+	// First Calculate how many pages are needed. One is always there.
+	pages = 1;
+	int gunIndex = 0;
+	int turretIndex = 0;
+	int rowsPerPage = static_cast<int>((sprite->Height() * scale) / 10.) - 1;
+	int offset = 0;
+	while(pageNeeded)
+	{
+		if(gunIndex < gunRows)
+		{
+			if((gunRows - gunIndex) * 20. >= height)
+				gunIndex += (rowsPerPage / 2);
+			else
+			{
+				offset = gunRows - gunIndex;
+				gunIndex = gunRows;
+			}
+		}
+		if(gunIndex >= gunRows && turretIndex < turretRows)
+		{
+			if(offset)
+				if((turretRows - turretIndex + offset) * 20. + 10. >= height)
+					turretIndex += (rowsPerPage / 2) - (offset + 1);
+				else
+					turretIndex = turretRows;
+			else
+			{
+				if((turretRows - turretIndex) * 20. >= height)
+					turretIndex += (rowsPerPage / 2);
+				else
+					pageNeeded = false;
+			}
+		}
+		if(turretIndex >= turretRows && gunIndex >= gunRows)
+			pageNeeded = false;
+		pages++;
+		offset = 0;
+	}
+	weaponsRight.clear();
+	weaponsLeft.clear();
+	indecesRight.clear();
+	indecesLeft.clear();
+	int index = 0;
 	for(const Hardpoint &hardpoint : ship.Weapons())
 	{
-		string name = "[empty]";
-		if(hardpoint.GetOutfit())
-			name = hardpoint.GetOutfit()->Name();
-
-		bool isRight = (hardpoint.GetPoint().X() >= 0.);
-		bool isTurret = hardpoint.IsTurret();
-
-		double &y = nextY[isRight][isTurret];
-		double x = centerX + (isRight ? LABEL_DX : -LABEL_DX - LABEL_WIDTH);
-		bool isHover = (index == hoverIndex);
-		layout.align = isRight ? Alignment::LEFT : Alignment::RIGHT;
-		font.Draw({name, layout}, Point(x, y + TEXT_OFF), isHover ? bright : dim);
-		Point zoneCenter(labelCenter[isRight], y + .5 * LINE_HEIGHT);
-		zones.emplace_back(zoneCenter, LINE_SIZE, index);
-
-		// Determine what color to use for the line.
-		float high = (index == hoverIndex ? .8f : .5f);
-		Color color(high, .75f * high, 0.f, 1.f);
-		if(isTurret)
-			color = Color(0.f, .75f * high, high, 1.f);
-
-		// Draw the line.
-		Point from(fromX[isRight], zoneCenter.Y());
-		Point to = bounds.Center() + (2. * scale) * hardpoint.GetPoint();
-		DrawLine(from, to, color);
-		if(isHover)
+		if(hardpoint.GetPoint().X() >= 0.)
+		{
+			indecesRight.emplace_back(index);
+			weaponsRight.emplace_back(&hardpoint);
+		}
+		else
+		{
+			indecesLeft.emplace_back(index);
+			weaponsLeft.emplace_back(&hardpoint);
+		}
+		index++;
+	}
+	for(int weaponIndex = (pageIndex - 1) * (rowsPerPage / 2); weaponIndex < pageIndex * (rowsPerPage / 2); weaponIndex++)
+	{
+		if(weaponsRight.size() > static_cast<unsigned int>(weaponIndex))
 		{
-			topFrom = from;
-			topTo = to;
-			topColor = color;
-			hasTop = true;
+			const Hardpoint *hardpoint = weaponsRight[weaponIndex];
```

Comment:
This code feels like its repeated twice (or almost), it should probably be made into a method, although I'd prefer a lambda function (but those can be tricky to use so do say if you need help)
Such method would take a bool to know if its left or right that we're drawing (which, why do you separate the two btw? it seemed fine to me with the x allowing to know if its left or right)

Reply:
Well the lambda can just take the vector as an argument

Reply:
Added a lambda as requested

Reply:
Nice! You could always select the elements that are to make it trough and which dont (instead of writing & only writing those you need)
If there are a lot of elements its fine this way for me